### PR TITLE
Add bow enchantments

### PR DESCRIPTION
**Pull request by JustTalDevelops**
Adds flame, infinity, power, and punch enchantments.

### CODE REVIEW

File: server/entity/arrow.go [hunk 1]
```diff
@@ -31,24 +32,27 @@ type Arrow struct {
...
 // NewArrow creates a new Arrow and returns it. It is equivalent to calling NewTippedArrow with `potion.Potion{}` as
 // tip.
-func NewArrow(pos mgl64.Vec3, yaw, pitch float64, owner world.Entity) *Arrow {
-	return NewTippedArrow(pos, yaw, pitch, owner, potion.Potion{})
+func NewArrow(pos mgl64.Vec3, yaw, pitch, damage float64, owner world.Entity) *Arrow {
+	return NewTippedArrow(pos, yaw, pitch, damage, owner, potion.Potion{})
 }
 
 // NewTippedArrow creates a new Arrow with a potion effect added to an entity when hit. The Arrow's base damage may be
 // changed by calling Arrow.SetBaseDamage. It is 2.0 by default.
```
Comment (Sandertv) [inline server/entity/arrow.go:50]: This documentation here isn't right anymore. Might want to correct this too.

File: server/entity/arrow.go [hunk 2]
```diff
@@ -222,10 +260,11 @@ func (a *Arrow) ignores(entity world.Entity) bool {
 }
 
 // New creates and returns an Arrow with the position, velocity, yaw, and pitch provided. It doesn't spawn the Arrow
-// by itself.
-func (a *Arrow) New(pos, vel mgl64.Vec3, yaw, pitch float64, owner world.Entity, critical, disallowPickup, obtainArrowOnPickup bool, tip potion.Potion) world.Entity {
-	arrow := NewTippedArrow(pos, yaw, pitch, owner, tip)
+// by itself. The second return value is true if an Arrow should be consumed.
```
Comment (Sandertv) [inline server/entity/arrow.go:263]: This documentation isn't correct anymore either.

**Review by Sandertv — CHANGES_REQUESTED** [review]
Most of it looks good to me. I've left a couple of comments. If someone else could have another look at the actual calculations and logic to make sure they match vanilla bedrock edition, that would be appreciated.

File: server/entity/arrow.go [hunk 3]
```diff
@@ -31,24 +32,27 @@ type Arrow struct {
...
-func NewArrow(pos mgl64.Vec3, yaw, pitch float64, owner world.Entity) *Arrow {
-	return NewTippedArrow(pos, yaw, pitch, owner, potion.Potion{})
+func NewArrow(pos mgl64.Vec3, yaw, pitch, damage float64, owner world.Entity) *Arrow {
+	return NewTippedArrow(pos, yaw, pitch, damage, owner, potion.Potion{})
 }
 
 // NewTippedArrow creates a new Arrow with a potion effect added to an entity when hit. The Arrow's base damage may be
 // changed by calling Arrow.SetBaseDamage. It is 2.0 by default.
-func NewTippedArrow(pos mgl64.Vec3, yaw, pitch float64, owner world.Entity, tip potion.Potion) *Arrow {
+func NewTippedArrow(pos mgl64.Vec3, yaw, pitch, damage float64, owner world.Entity, tip potion.Potion) *Arrow {
```
[the remark this answers is not in the document]
Comment (JustTalDevelops) [reply server/entity/arrow.go]: Created those functions.

File: server/entity/arrow.go (hunk 1 above)
Comment (JustTalDevelops) [reply server/entity/arrow.go]: Correct.d

File: server/entity/arrow.go (hunk 2 above)
Comment (JustTalDevelops) [reply server/entity/arrow.go]: Removed.

File: server/entity/arrow.go [hunk 4]
```diff
@@ -204,14 +228,28 @@ func (a *Arrow) Tick(w *world.World, current int64) {
...
 				}
+				if flammable, ok := living.(Flammable); ok && a.OnFireDuration() > 0 {
+					flammable.SetOnFire(time.Second * 5)
+				}
+				if a.punchLevel > 0 {
+					horizontalVel := pastVel
+					horizontalVel[1] = 0
+					if speed := horizontalVel.Len(); speed > 0 {
+						multiplier := (enchantment.Punch{}).PunchMultiplier(a.punchLevel, speed)
+						living.SetVelocity(living.Velocity().Add(mgl64.Vec3{pastVel[0] * multiplier, 0.1, pastVel[2] * multiplier}))
```
[the remark this answers is not in the document]
Comment (JustTalDevelops) [reply server/entity/arrow.go]: Yeah that's true, changed.

File: server/entity/living.go [hunk 5]
```diff
@@ -38,6 +38,10 @@ type Living interface {
 	// source of the velocity, typically the position of an attacking entity. The source is used to calculate
 	// the direction which the entity should be knocked back in.
 	KnockBack(src mgl64.Vec3, force, height float64)
+	// Velocity returns the players current velocity.
+	Velocity() mgl64.Vec3
+	// SetVelocity updates the entity's velocity.
+	SetVelocity(velocity mgl64.Vec3)
```
[the remark this answers is not in the document]
Comment (JustTalDevelops) [reply server/entity/living.go]: Removed.

**Review by DaPigGuy — CHANGES_REQUESTED** [review]
Several big issues to fix,

* Punch does not have any impact whatsoever, looks like `force` & `height` are not used?
* Damage done by bows w/o Power are not consistent between df & Minecraft, and the Power enchantment damage bonuses also don't appear to be consistent with what the wiki has to say.
  * Tests done on Hard difficulty with an unenchanted Bow & Power III Bow. In vanilla Minecraft, the bows are consistently doing 4.5/9 hearts of damage respectively while df seems to do a varying amount of damage

**Comment by JustTalDevelops** [comment]
Did you test with critical arrows? Critical arrows are meant to do a varying amount of extra, inconsistent damage.

**Comment by JustTalDevelops** [comment]
Seems like this isn't the case on Bedrock, instead they seem to have a `1.8` damage multiplier. Will try this and see if I can reproduce the results you're getting.

**Comment by JustTalDevelops** [comment]
I believe the damage values should now match Bedrock @DaPigGuy.

File: server/entity/arrow.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (JustTalDevelops) [reply server/entity/arrow.go:256]: After some testing, this actually does seem to be correct and is producing values that are accurate to BDS. I think it's best we leave this here as long as BDS does the same.

**Comment by JustTalDevelops** [comment]
Okay so after some testing, arrow damage seems to be fairly accurate but it still does not match vanilla.

I think that's out of the scope of this PR though, as the actual calculated power damage is fine, the issues just arise with arrow damage in general. Should probably open a separate issue for that, but the enchantments themselves seem to work flawlessly.

**Review by Sandertv — APPROVED** [review]
If this has all been tested, I think it's ready to be merged. Feel free to merge at your own discretion.

**Review by Sandertv — CHANGES_REQUESTED** [review]
Now that TNT has been merged, flame arrows need to ignite TNT.

**Pull request merged**